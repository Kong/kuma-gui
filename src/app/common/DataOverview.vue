<template>
  <div
    class="data-overview"
    data-testid="data-overview"
  >
    <div class="data-table-controls mb-2">
      <slot name="additionalControls" />

      <KButton
        class="refresh-button"
        appearance="primary"
        :disabled="isLoading"
        icon="redo"
        data-testid="data-overview-refresh-button"
        @click="onRefreshButtonClick"
      >
        Refresh
      </KButton>
    </div>

    <LoadingBlock v-if="isLoading" />

    <ErrorBlock
      v-else-if="error !== null"
      :error="error"
    />

    <EmptyBlock v-else-if="isEmpty" />

    <div
      v-else
      class="data-overview-content"
    >
      <!-- data -->
      <div
        v-if="!tableDataIsEmpty && tableData"
        class="data-overview-table"
      >
        <KTable
          :key="tableRecompuationKey"
          class="data-overview-table"
          :class="{ 'data-table-is-hidden' : tableDataIsEmpty }"
          :fetcher="tableDataFetcher"
          :headers="tableHeaders"
          :cell-attrs="getCellAttributes"
          :row-attrs="getRowAttributes"
          disable-pagination
          is-clickable
          data-testid="data-overview-table"
          @row:click="tableRowHandler"
        >
          <!-- Custom slots provided by parent -->
          <template
            v-for="slot in customSlots"
            #[slot]="{ rowValue, row }"
          >
            <slot
              :name="slot"
              :row-value="rowValue"
              :row="row"
            />
          </template>

          <!-- status -->
          <template #status="{ rowValue }">
            <StatusBadge
              v-if="rowValue"
              :status="rowValue"
            />

            <template v-else>
              —
            </template>
          </template>

          <!-- tags -->
          <template #tags="{ rowValue }">
            <TagList :tags="rowValue" />
          </template>

          <template #name="{ row, rowValue }">
            <router-link
              v-if="row.nameRoute"
              :to="row.nameRoute"
            >
              {{ rowValue }}
            </router-link>

            <template v-else>
              {{ rowValue }}
            </template>
          </template>

          <template #mesh="{ row, rowValue }">
            <router-link
              v-if="row.meshRoute"
              :to="row.meshRoute"
            >
              {{ rowValue }}
            </router-link>

            <template v-else>
              {{ rowValue }}
            </template>
          </template>

          <template #service="{ row, rowValue }">
            <router-link
              v-if="row.serviceInsightRoute"
              :to="row.serviceInsightRoute"
            >
              {{ rowValue }}
            </router-link>

            <template v-else>
              {{ rowValue }}
            </template>
          </template>

          <template #zone="{ row, rowValue }">
            <router-link
              v-if="row.zoneRoute"
              :to="row.zoneRoute"
            >
              {{ rowValue }}
            </router-link>

            <template v-else>
              {{ rowValue }}
            </template>
          </template>

          <!--- total Updates --->
          <template #totalUpdates="{ row }">
            <span>
              {{ row.totalUpdates }}
            </span>
          </template>

          <!--- dp Version --->
          <template #dpVersion="{ row, rowValue }">
            <div
              :class="{
                'with-warnings': row.unsupportedEnvoyVersion || row.unsupportedKumaDPVersion || row.kumaDpAndKumaCpMismatch,
              }"
            >
              {{ rowValue }}
            </div>
          </template>

          <!--- envoy Version --->
          <template #envoyVersion="{ row, rowValue }">
            <div
              :class="{
                'with-warnings': row.unsupportedEnvoyVersion,
              }"
            >
              {{ rowValue }}
            </div>
          </template>

          <!--- warnings --->
          <template
            v-if="showWarnings"
            #warnings="{ row }"
          >
            <KIcon
              v-if="row.withWarnings"
              class="mr-1"
              icon="warning"
              color="var(--black-75)"
              secondary-color="var(--yellow-300)"
              size="20"
            />
          </template>

          <template
            v-if="showDetails"
            #details="{ row }"
          >
            <KButton
              class="detail-link"
              appearance="btn-link"
              :to="row.nameRoute"
            >
              <template #icon>
                <KIcon
                  :icon="row.warnings.length > 0 ? 'warning' : 'info'"
                  :color="row.warnings.length > 0 ? 'var(--black-75)' : 'var(--blue-500)'"
                  :secondary-color="row.warnings.length > 0 ? 'var(--yellow-300)' : undefined"
                  size="16"
                  hide-title
                />
              </template>
              Details
            </KButton>
          </template>
        </KTable>

        <PaginationWidget
          :has-previous="internalPageOffset > 0"
          :has-next="Boolean(next)"
          @next="goToNextPage"
          @previous="goToPreviousPage"
        />
      </div>

      <EmptyBlock v-if="tableDataIsEmpty && tableData">
        <template #title>
          <img
            class="mb-3"
            src="@/assets/images/icon-empty-table.svg?url"
          >

          <p v-if="emptyState.title">
            {{ emptyState.title }}
          </p>

          <p v-else>
            No items found
          </p>
        </template>

        <template
          v-if="emptyState.message"
          #message
        >
          {{ emptyState.message }}
        </template>
      </EmptyBlock>

      <div
        v-if="$slots.content"
        class="data-overview-content mt-6"
      >
        <slot name="content" />
      </div>
    </div>
  </div>
</template>

<script lang="ts" setup>
import { computed, PropType, ref, useSlots, watch } from 'vue'
import { KButton, KIcon, KTable } from '@kong/kongponents'
import { datadogLogs } from '@datadog/browser-logs'

import { datadogLogEvents } from '@/utilities/datadogLogEvents'
import EmptyBlock from './EmptyBlock.vue'
import ErrorBlock from './ErrorBlock.vue'
import LoadingBlock from './LoadingBlock.vue'
import PaginationWidget from './PaginationWidget.vue'
import StatusBadge from '@/app/common/StatusBadge.vue'
import TagList from './TagList.vue'
import { TableData } from '@/types/index.d'

const slots = useSlots()

const props = defineProps({
  selectedEntityName: {
    type: String,
    required: false,
    default: '',
  },

  pageSize: {
    type: Number,
    required: false,
    default: 12,
  },

  isLoading: {
    type: Boolean,
    required: false,
    default: false,
  },

  error: {
    type: [Error, null] as PropType<Error | null>,
    required: false,
    default: null,
  },

  isEmpty: {
    type: Boolean,
    required: false,
    default: false,
  },

  emptyState: {
    type: Object,
    required: false,
    default: null,
  },

  tableData: {
    type: Object as PropType<TableData>,
    required: false,
    default: null,
  },

  tableDataIsEmpty: {
    type: Boolean,
    required: false,
    default: false,
  },

  showWarnings: {
    type: Boolean,
    required: false,
    default: false,
  },

  showDetails: {
    type: Boolean,
    required: false,
    default: false,
  },

  next: {
    type: [String, Boolean, null] as PropType<string | boolean | null>,
    required: false,
    default: false,
  },

  pageOffset: {
    type: Number,
    required: false,
    default: 0,
  },
})

const emit = defineEmits(['table-action', 'refresh', 'load-data'])

const selectedRow = ref('')
const internalPageOffset = ref(props.pageOffset)

const tableHeaders = computed(() => {
  if (!props.showWarnings) {
    return props.tableData.headers.filter((header) => header.key !== 'warnings')
  } else {
    return props.tableData.headers
  }
})
const customSlots = computed(() => props.tableData.headers.map((header) => header.key).filter((key) => slots[key]))
/**
 * Just changing the table data/headers doesn’t cause the table component to re-render, so by updating its `key` prop when this happens, we force a re-render. This is necessary to get the column visibility toggles and similar operations to work in the data planes list view.
 */
const tableRecompuationKey = computed(() => `${props.tableData.data.length}-${tableHeaders.value.length}`)

watch(() => props.isLoading, function () {
  if (!props.isLoading && props.tableData.data.length > 0) {
    selectedRow.value = props.selectedEntityName || props.tableData.data[0].name
  }
})

function tableDataFetcher(): { data: object[], total: number } {
  return {
    data: props.tableData.data,
    total: props.tableData.data.length,
  }
}

function tableRowHandler(_e: any, row: any): void {
  selectedRow.value = row.name
  emit('table-action', row)
}

function onRefreshButtonClick(): void {
  emit('refresh')
  emit('load-data', internalPageOffset.value)
  datadogLogs.logger.info(datadogLogEvents.TABLE_REFRESH_BUTTON_CLICKED)
}

function goToPreviousPage(): void {
  internalPageOffset.value = props.pageOffset - props.pageSize
  emit('load-data', props.pageOffset - props.pageSize)
}

function goToNextPage(): void {
  internalPageOffset.value = props.pageOffset + props.pageSize
  emit('load-data', props.pageOffset + props.pageSize)
}

function getCellAttributes({ headerKey }: any): Record<string, string> {
  const className = ['warnings'].includes(headerKey) ? 'text-center' : ['details'].includes(headerKey) ? 'text-right' : ''

  return { class: className }
}

function getRowAttributes({ name }: any): Record<string, string> {
  const entityName = props.selectedEntityName || props.tableData.data[0].name
  const className = name === entityName ? 'is-selected' : ''

  return { class: className }
}
</script>

<style lang="scss" scoped>
.refresh-icon {
  &.is-spinning g {
    animation: spin 1.2s infinite linear;
  }

  @keyframes spin {
    0% {
      transform: rotate(0deg);
    }

    100% {
      transform: rotate(1turn);
    }
  }
}

.data-table-controls {
  display: flex;
  justify-content: flex-end;
  align-items: center;
  gap: var(--spacing-md);
  padding: var(--spacing-sm) var(--spacing-sm) 0 var(--spacing-sm);
}

.with-warnings {
  color: var(--yellow-500);
}
</style>

<style lang="scss">
.data-overview-table .is-selected {
  background-color: var(--grey-100);
}
</style>
