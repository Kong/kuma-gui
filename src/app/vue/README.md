---
section: Packages
---
# vue

Automatic `vue` initialization using our projects dependency injection layer.

