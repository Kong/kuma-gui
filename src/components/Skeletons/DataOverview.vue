<template>
  <div class="data-overview">
    <!-- controls -->
    <div
      v-if="displayRefreshControl"
      class="data-table-controls mb-2"
    >
      <KButton
        appearance="primary"
        size="small"
        :disabled="isLoading"
        @click="$emit('reloadData')"
      >
        <KIcon
          v-if="isLoading"
          icon="spinner"
          color="#fff"
          size="48"
        />
        <span>Refresh</span>
      </KButton>
    </div>

    <div
      v-if="isReady"
      class="data-overview-content"
    >
      <!-- metrics -->
      <MetricGrid
        v-if="!isLoading && displayMetrics && metricsData"
        :metrics="metricsData"
      />

      <!-- data -->
      <div
        v-if="displayDataTable && !tableDataIsEmpty && tableData"
        class="data-overview-table"
      >
        <KTable
          class="micro-table"
          :class="{ 'data-table-is-hidden' : tableDataIsEmpty, 'has-border': tableHasBorder }"
          :options="tableDataFiltered"
          has-hover
          @row:click="tableRowHandler"
        >
          <!-- status -->
          <template
            v-if="displayTableDataStatus"
            v-slot:status="{ rowValue }"
          >
            <div
              class="entity-status"
              :class="{ 'is-offline': (rowValue.toLowerCase() === 'offline') }"
            >
              <span class="entity-status__dot" />
              <span class="entity-status__label">{{ rowValue }}</span>
            </div>
          </template>
          <!-- tags -->
          <template
            v-slot:tags="{ rowValue }"
          >
            <span
              v-for="(item, key) in rowValue"
              :key="key"
              class="entity-tags"
              :class="`entity-tags--${key}`"
            >
              <span
                class="entity-tags__label"
                :class="`entity-tags__label--${item.label.toLowerCase()}`"
              >
                {{ item.label }}
              </span>
              <span
                class="entity-tags__value"
                :class="`entity-tags__value--${item.value}`"
              >
                {{ item.value }}
              </span>
            </span>
          </template>

          <template
            slot="actions"
            slot-scope="{ row }"
          >
            <a
              v-if="tableDataFunctionText"
              class="data-table-action-link"
              :class="{ 'is-active': ($store.state.selectedTableRow === row.name) }"
            >
              <span
                v-if="$store.state.selectedTableRow === row.name"
                class="action-link__active-state"
              >
                &#x2713;
                <span class="sr-only">
                  Selected
                </span>
              </span>
              <span
                v-else
                class="action-link__normal-state"
              >
                {{ tableDataFunctionText }}
              </span>
            </a>
          </template>
        </KTable>

        <Pagination
          v-if="tableData && tableRowCount > pageSize"
          :has-previous="pageNumber > 0"
          :has-next="pageNumber < pageCount -1"
          @next="goToNextPage"
          @previous="goToPreviousPage"
        />
      </div>

      <!-- empty state if no items are found -->
      <KEmptyState
        v-if="displayDataTable && tableDataIsEmpty && tableData"
        cta-is-hidden
      >
        <template slot="title">
          <div class="card-icon mb-3">
            <img src="~@/assets/images/icon-empty-table.svg?external">
          </div>
          <span v-if="emptyState.title">
            {{ emptyState.title }}
          </span>
          <span v-else>
            No Items Found
          </span>
        </template>
        <template
          v-if="emptyState.message"
          slot="message"
        >
          {{ emptyState.message }}
        </template>
      </KEmptyState>

      <!-- additional page content -->
      <div
        v-if="$slots.content"
        class="data-overview-content mt-6"
      >
        <slot name="content" />
      </div>
    </div>

    <!-- loading state -->
    <KEmptyState
      v-if="isLoading"
      cta-is-hidden
    >
      <template slot="title">
        <div class="card-icon mb-3">
          <KIcon
            icon="spinner"
            color="rgba(0, 0, 0, 0.1)"
            size="42"
          />
        </div>
        Data Loading...
      </template>
    </KEmptyState>

    <!-- error has occurred -->
    <KEmptyState
      v-if="hasError"
      cta-is-hidden
    >
      <template slot="title">
        <div class="card-icon mb-3">
          <KIcon
            class="kong-icon--centered"
            color="var(--yellow-base)"
            icon="warning"
            size="42"
          />
        </div>
        An error has occurred while trying to load this data.
      </template>
    </KEmptyState>
  </div>
</template>

<script>
import MetricGrid from '@/components/Metrics/MetricGrid'
import Pagination from '@/components/Pagination'

export default {
  name: 'DataOverview',
  components: {
    MetricGrid,
    Pagination
  },
  props: {
    pageSize: {
      type: Number,
      default: 12
    },
    displayMetrics: {
      type: Boolean,
      default: false
    },
    metricsData: {
      type: Array,
      default: null
    },
    isLoading: {
      type: Boolean,
      default: false
    },
    hasError: {
      type: Boolean,
      default: false
    },
    isEmpty: {
      type: Boolean,
      default: false
    },
    emptyState: {
      type: Object,
      default: null
    },
    ctaAction: {
      type: Object,
      default: () => {}
    },
    showCta: {
      type: Boolean,
      default: true
    },
    displayDataTable: {
      type: Boolean,
      default: false
    },
    tableData: {
      type: Object,
      default: null
    },
    tableHasBorder: {
      type: Boolean,
      required: false,
      default: false
    },
    tableDataIsEmpty: {
      type: Boolean,
      default: false
    },
    tableDataActionsLink: {
      type: String,
      default: null
    },
    tableActionsRouteName: {
      type: String,
      default: null
    },
    displayTableDataStatus: {
      type: Boolean,
      default: true
    },
    displayRefreshControl: {
      type: Boolean,
      default: true
    },
    tableDataRow: {
      type: String,
      required: false,
      default: 'name'
    },
    tableDataFunctionText: {
      type: String,
      required: false,
      default: null
    }
  },
  data () {
    return {
      pageNumber: 0
    }
  },
  computed: {
    isReady () {
      return !this.isEmpty && !this.hasError && !this.isLoading
    },
    tableRowCount () {
      return Object.entries(this.tableData.data).length
    },
    pageCount () {
      const itemCount = Object.entries(this.tableData.data).length
      const pageSize = this.pageSize

      return Math.ceil(itemCount / pageSize)
    },
    tableDataFiltered () {
      const data = this.tableData.data
      const headers = this.tableData.headers
      const start = this.pageNumber * this.pageSize
      const end = start + this.pageSize
      const filtered = data.slice(start, end)
      const newData = { headers, data: [...filtered] }

      return newData
    }
  },
  methods: {
    goToPreviousPage () {
      this.pageNumber--
    },
    goToNextPage () {
      this.pageNumber++
    },
    tableRowHandler (e, row, type) {
      this.$emit('tableAction', row.name)
    }
  }
}
</script>

<style lang="scss">
.empty-state-title {

  .card-icon {
    text-align: center;

    img, svg {
      display: block;
      margin-left: auto;
      margin-right: auto;
    }
  }
}

.data-table-controls {
  text-align: right;
  padding: var(--spacing-sm) var(--spacing-sm) 0 var(--spacing-sm);

  button:after {
    display: none;
  }
}

.info-grid {

  .metric {
    margin-bottom: 16px;
  }
}

.empty-state-wrapper {
  margin-bottom: 2em;
}

.entity-tags {
  display: inline-flex;
  align-items: stretch;
  font-size: 12px;
  background-color: #fff;
  font-family: var(--font-family-mono);

  &:not(:last-of-type) {
    margin-right: 0.5rem;
    margin-bottom: 0.25rem;
  }
}

.entity-tags__label {
  --color: var(--blue-1);

  position: relative;
  background-color: var(--color);
  color: #fff;
  text-transform: uppercase;
  border-radius: 5px 0 0 5px;
  padding: 0.15rem 0.5rem;
  box-shadow: inset 0 0 0 1px var(--color);
}

.entity-tags__label--service,
.entity-tags__label--protocol {
  --color: var(--brand-color-6);

  background-color: var(--color);
  box-shadow: inset 0 0 0 1px var(--color);
}

.entity-tags__value {
  background-color: #fff;
  border-radius: 0 5px 5px 0;
  padding: 0.15rem 0.5rem 0.15rem 0.75rem;
  color: currentColor;
  box-shadow: inset 0 0 0 1px currentColor;
}

.data-overview-table {

}

.micro-table.micro-table {
  --dp-table-font-size: 14px;
  --dp-table-padding: 10px;

  th, td {
    padding: var(--dp-table-padding);
  }
}

.k-table {
  font-size: var(--dp-table-font-size);

  tr {
    position: relative;
    overflow: hidden;
    cursor: pointer;
  }

  th {
    background-color: var(--gray-7);
  }

  thead {
    border-top: 0 !important;
    border-bottom-width: 1px !important;
  }

  &.has-border {
    border: 1px solid var(--gray-4);
    border-bottom: 0;
  }

  .data-table-action-link {
    // position: absolute;
    // top: 0; right: 0; bottom: 0; left: 0;
    display: block;
    text-align: right;
    padding: var(--spacing-sm);
    cursor: pointer;
    overflow: hidden;
    padding: 0;

    &.is-active {
      display: flex;
      align-items: center;
      justify-content: flex-end;
      // position: absolute;
      // top: 0; left: 0;
      width: 100%;
      height: 100%;

      &:before {
        // position: absolute;
        // top: 0; left: 0;
        // z-index: -1;
        // width: 100%;
        // height: 100%;
        // display: block;
        // content: "";
        // background-color: var(--blue-lightest);
        // text-align: center;
      }
    }
  }

  .action-link__active-state {
    --size: 18px;

    // position: absolute;
    text-align: right;
    display: block;
    width: var(--size);
    height: var(--size);
    line-height: var(--size);
    border-radius: 50%;
    background-color: var(--logo-green);
    margin: 0 5px 0 auto;
    color: #fff;
    font-size: 13px;
    text-align: center;

    &:before {
      display: block;
    }
  }

}

@media only screen and (min-width: 841px) {
  .info-grid {
    flex-wrap: wrap;

    .metric {
      flex: 0 0 calc(33.33% - 32px); // gutter = 16px
    }
  }
}
</style>
