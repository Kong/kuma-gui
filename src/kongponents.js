import Vue from 'vue'
import '@kongponents/styles/styles.css'
import KAlert from '@kongponents/kalert'
import KBadge from '@kongponents/kbadge'
import KButton from '@kongponents/kbutton'
import KCard from '@kongponents/kcard'
import KEmptyState from '@kongponents/kemptystate'
import KIcon from '@kongponents/kicon'
import KTable from '@kongponents/ktable'
import KToaster from '@kongponents/ktoaster'
import KoolTip from '@kongponents/kooltip'
import KClipboardProvider from '@kongponents/kclipboardprovider'
import Krumbs from '@kongponents/krumbs'
import KPop from '@kongponents/kpop'
import KToggle from '@kongponents/ktoggle'
import KTabs from '@kongponents/ktabs'
import KModal from '@kongponents/kmodal'
import KInput from '@kongponents/kinput'
import KInputSwitch from '@kongponents/kinputswitch'

Vue.component('KAlert', KAlert)
Vue.component('KBadge', KBadge)
Vue.component('KButton', KButton)
Vue.component('KCard', KCard)
Vue.component('KEmptyState', KEmptyState)
Vue.component('KIcon', KIcon)
Vue.component('KTable', KTable)
Vue.component('KToaster', KToaster)
Vue.component('KoolTip', KoolTip)
Vue.component('KClipboardProvider', KClipboardProvider)
Vue.component('Krumbs', Krumbs)
Vue.component('KPop', KPop)
Vue.component('KTabs', KTabs)
Vue.component('KToggle', KToggle)
Vue.component('KInput', KInput)
Vue.component('KInputSwitch', KInputSwitch)
Vue.component('KModal', KModal)
