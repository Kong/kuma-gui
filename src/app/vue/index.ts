import { createApp } from 'vue'
import {
  createRouter,
  createWebHistory,
} from 'vue-router'

import type { ServiceDefinition } from '@/services/utils'
import { token, createInjections } from '@/services/utils'
import type { Component } from 'vue'
import type { Router, RouteRecordRaw, NavigationGuard } from 'vue-router'

type Token = ReturnType<typeof token>

type VueApp = ReturnType<typeof createApp>
export type PluginDefinition = Parameters<VueApp['use']>
export type ComponentDefinition = [string, Component]

const $ = {
  app: token<(App: Component) => Promise<VueApp>>('vue.app'),
  router: token<Router>('vue.router'),

  components: token('vue.components'),
  plugins: token<PluginDefinition[]>('vue.plugins'),
  routes: token<RouteRecordRaw[]>('vue.routes'),
  routesLabel: token<RouteRecordRaw[]>('vue.routes.label'),
  navigationGuards: token<NavigationGuard[]>('vue.routes.navigation.guards'),
}
export const services = (app: Record<string, Token>): ServiceDefinition[] => {
  return [
    [$.app, {
      service: (
        components: ComponentDefinition[],
        plugins: PluginDefinition[],
      ) => {
        return async (App: Component) => {
          const app = createApp(App)

          plugins.forEach(([...args]) => {
            app.use(...args)
          })

          components.forEach(([name, component]: [string, Component]) => {
            app.component(name, component)
          })

          return app
        }
      },
      arguments: [
        $.components,
        $.plugins,
      ],
    }],
    [$.router, {
      service: (
        env: (str: 'KUMA_BASE_PATH') => string,
        routes: RouteRecordRaw[],
        guards: NavigationGuard[],
      ) => {
        const router = createRouter({
          history: createWebHistory(env('KUMA_BASE_PATH')),
          routes,
        })

        guards.forEach((item) => {
          if (typeof item === 'function') {
            router.beforeEach(item)
          }
        })

        return router
      },
      arguments: [
        app.env,
        $.routes,
        $.navigationGuards,
      ],
    }],

    [token('vue.plugins'), {
      service: (router) => {
        return [
          [router],
        ]
      },
      arguments: [
        $.router,
      ],
      labels: [
        $.plugins,
      ],
    }],

    // TODO(jc): Label decoration. Make it so we can decorate labels
    // Temporarily turn the routesLabel label into routes for decoration
    // purposes
    [$.routes, {
      service: (routes: RouteRecordRaw[]) => routes,
      arguments: [
        $.routesLabel,
      ],
    }],
    //

    // TODO(jc): Make it so we don't need to provide a [] empty default for
    // labels
    [token('application.routes.navigation.guards'), {
      service: () => {
        return []
      },
      labels: [
        $.navigationGuards,
      ],
    }],
  ]
}

export const TOKENS = $
export const [
  useRouter,
] = createInjections(
  TOKENS.router,
)
