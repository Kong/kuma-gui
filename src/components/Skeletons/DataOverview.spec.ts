import { flushPromises, RouterLinkStub } from '@vue/test-utils'
import { render, screen } from '@testing-library/vue'
import { datadogLogs } from '@datadog/browser-logs'
import userEvent from '@testing-library/user-event'
import { KButton, KEmptyState, KIcon, KTable } from '@kong/kongponents'

import DataOverview from './DataOverview.vue'

jest.mock('@datadog/browser-logs')

function renderComponent(props = {}) {
  return render(DataOverview, {
    global: {
      components: {
        KButton,
        KEmptyState,
        KIcon,
        KTable,
      },
      stubs: {
        'router-link': RouterLinkStub,
      },
    },
    props,
    slots: {
      custom: `
        <ul>
          <li>test</li>
        </ul>
      `,
    },
  })
}

describe('DataOverview.vue', () => {
  beforeEach(() => {
    (datadogLogs.logger.info as jest.MockedFunction<any>).mockClear()
  })

  it('renders basic snapshot', () => {
    const { container } = renderComponent({
      tableData: {
        headers: [],
        data: [],
      },
    })

    expect(container).toMatchSnapshot()
  })

  it('renders additional scoped slot', async () => {
    renderComponent({
      tableData: {
        headers: [
          { key: 'custom', hideLabel: true },
        ],
        data: [
          {
            custom: ['custom', 'values', 'in', 'an', 'array'],
          },
        ],
      },
    })

    await flushPromises()

    expect(screen.getByRole('table')).toMatchSnapshot()
  })

  it('renders pagination and react on click', async () => {
    renderComponent({
      next: true,
      tableData: {
        headers: [],
        data: [],
      },
    })

    const next = screen.getByText(/Next/)

    expect(next).toBeInTheDocument()
    expect(screen.queryByText(/Previous/)).not.toBeInTheDocument()

    await userEvent.click(next)

    const back = screen.queryByText(/Previous/)

    expect(next).toBeInTheDocument()
    expect(back).toBeInTheDocument()

    await userEvent.click(back as HTMLElement)

    expect(next).toBeInTheDocument()
    expect(back).not.toBeInTheDocument()
  })

  it('refresh page on second page', async () => {
    renderComponent({
      next: true,
      tableData: {
        headers: [],
        data: [],
      },
    })

    const next = screen.getByText(/Next/)
    const refresh = screen.getByText(/Refresh/)

    expect(next).toBeInTheDocument()

    await userEvent.click(next)
    await userEvent.click(refresh)

    expect(datadogLogs.logger.info).toMatchSnapshot()
  })

  it('renders all custom templates for data', async () => {
    const { container } = renderComponent({
      showWarnings: true,
      tableData: {
        headers: [
          { key: 'actions', hideLabel: true },
          { label: 'Status', key: 'status' },
          { label: 'Total Updates', key: 'totalUpdates' },
          { label: 'Kuma DP version', key: 'dpVersion' },
          { label: 'Envoy version', key: 'envoyVersion' },
          { key: 'warnings', hideLabel: true },
        ],
        data: [
          {
            status: 'offline',
            totalUpdates: 1,
            dpVersion: 'foo',
            envoyVersion: '1.2',
            withWarnings: true,
          },
        ],
      },
    })

    await flushPromises()

    expect(container).toMatchSnapshot()
  })
})
