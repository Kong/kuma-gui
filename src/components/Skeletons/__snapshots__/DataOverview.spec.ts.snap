// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataOverview.vue refresh page on second page 1`] = `
Object {
  "loadData": Array [
    Array [
      12,
    ],
    Array [
      12,
    ],
  ],
}
`;

exports[`DataOverview.vue refresh page on second page 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "pagination-next-button-clicked",
    ],
    Array [
      "table-refresh-button-clicked",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`DataOverview.vue renders basic snapshot 1`] = `
<div>
  <div
    class="data-overview"
  >
    <div
      class="data-table-controls mb-2"
    >
       
      <button
        class="k-button ml-2 refresh-button small primary"
        data-v-90a15d32=""
        type="button"
      >
        <div
          class="refresh-icon"
          data-v-90a15d32=""
        >
          <svg
            data-v-90a15d32=""
            viewBox="0 0 36 36"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              data-v-90a15d32=""
              fill="#fff"
              fill-rule="nonzero"
            >
              <path
                d="M18 5.5a12.465 12.465 0 00-8.118 2.995 1.5 1.5 0 001.847 2.363l.115-.095A9.437 9.437 0 0118 8.5l.272.004a9.487 9.487 0 019.07 7.75l.04.246H25a.5.5 0 00-.416.777l4 6a.5.5 0 00.832 0l4-6 .04-.072A.5.5 0 0033 16.5h-2.601l-.017-.15C29.567 10.2 24.294 5.5 18 5.5zM2.584 18.723l-.04.072A.5.5 0 003 19.5h2.6l.018.15C6.433 25.8 11.706 30.5 18 30.5c3.013 0 5.873-1.076 8.118-2.995a1.5 1.5 0 00-1.847-2.363l-.115.095A9.437 9.437 0 0118 27.5l-.272-.004a9.487 9.487 0 01-9.07-7.75l-.041-.246H11a.5.5 0 00.416-.777l-4-6a.5.5 0 00-.832 0l-4 6z"
                data-v-90a15d32=""
              />
            </g>
          </svg>
        </div>
         
        <span
          data-v-90a15d32=""
        >
          Refresh
        </span>
      </button>
    </div>
     
    <div
      class="data-overview-content"
    >
      <!---->
       
      <div
        class="data-overview-table"
      >
        <table
          class="k-table micro-table has-hover is-clickable"
          data-v-59ad4099=""
        >
          <thead
            data-v-59ad4099=""
          >
            <tr
              data-v-59ad4099=""
            />
          </thead>
          <tbody
            data-v-59ad4099=""
          />
        </table>
         
        <div
          class="pagination"
        >
          <!---->
           
          <!---->
        </div>
      </div>
       
      <!---->
       
      <!---->
    </div>
     
    <!---->
     
    <!---->
  </div>
</div>
`;
